compiler/compile/OMRCompilation.cpp

```diff
@@ -735,38 +744,85 @@ bool OMR::Compilation::isPotentialOSRPointWithSupport(TR::TreeTop *tt)
    return potentialOSRPoint;
    }
 
+/*
+ * OSR can operate in two modes, voluntary and involuntary.
+ *
+ * In involuntary OSR, the JITed code does not control when an OSR transition occurs. It can be
+ * initiated externally at any potential OSR point.
+ *
+ * In voluntary OSR, the JITed code does control when an OSR transition occurs, allowing it to
+ * limit the OSR points with transitions.
+ */
+TR::OSRMode
+OMR::Compilation::getOSRMode()
+   {
+   if (self()->getOption(TR_FullSpeedDebug))
+      return TR::involuntaryOSR;
+   return TR::voluntaryOSR;
+   }
+
+/*
+ * The OSR transition destination may be before or after the OSR point.
+ * When located before, the transition will target the bytecode index of
+ * the OSR point, whilst those located after may have an offset bytecode
+ * index.
+ */
+TR::OSRTransitionTarget
+OMR::Compilation::getOSRTransitionTarget()
+   {
+   if (self()->getHCRMode() == TR::osr)
+      return TR::postExecutionOSR;
+   return TR::preExecutionOSR;
+   }
+
+/*
+ * Provides the bytecode offset between the OSR point and the destination
+ * of the transition. Only used when doing postExecutionOSR to indicate
+ * the appropriate bytecode index after the OSR point.
+ */
 int32_t
 OMR::Compilation::getOSRInductionOffset(TR::Node *node)
    {
-   if (self()->getHCRMode() == TR::osr)
+   // If no induction after the OSR point, offset must be 0
+   if (self()->getOSRTransitionTarget() != TR::postExecutionOSR)
+      return 0;
+
+   switch (node->getOpCodeValue())
       {
-      switch (node->getOpCodeValue())
-         {
-         case TR::monent: return 1;
-         case TR::asynccheck: return 0;
-         default: return 3;
-         }
+      case TR::monent: return 1;
+      case TR::asynccheck: return 0;
+      default: return 3;
```

Comment:
That is a good point - this should be case case case and default should assert that you have an unexpected opcode. @ncough FYI since one of us should fix this.